# Statistical Tools 
This Tools helps us to again on the same line in order make dicision / in order to conclude something .I want to do some type of test.Like Hypothesis Testing.
## 1 - Ttest or Student Ttest.
## 2 - Anova -Analysis of Variance
## 3 -Chi-Square

# 1-> Ttest or Student Test

There are Two types of Ttest 

- One Sample Test
- Two Sample Test

In [5]:
#Conduct ttest when standard deviation is not known.
import pandas as pd
hyp=pd.read_csv("https://raw.githubusercontent.com/training-ml/Files/main/Hypothesis%20ttest.csv")
hyp.head()

ID  Gender  Race  SEB  School  Prog  Read  Write  Math1  Math2  SST
0   70       0     4    1       1     1    57     52     41     47   57
1  121       1     4    2       1     3    68     59     53     63   61
2   86       0     4    3       1     1    44     33     54     58   31
3  141       0     4    3       1     3    63     44     47     53   56
4  172       0     4    2       1     2    47     52     57     53   61

A teacher wants to teach student for a competative exam, she collected sample of the student marks from different school and based on her past experience she believes that, these type of student will score more than 60 marks in the exam. Now, she wants to conduct hypothesis test on the marks.

Ho=mu>=60

Ha=mu<=60

## One Sample Test

It is also called as biased sample test because it focuses on only one sample and tell us the probability

In [12]:
xbar=hyp['Math1'].mean()
alpha=0.05
mu=60
import scipy.stats as st

#code is for one sample test is  - st.ttest_1samp(sample,mu)
st.ttest_1samp(hyp.Math1,mu)

TtestResult(statistic=-11.102736525713398, pvalue=1.3339158110547646e-22, df=199)

### Actual Value:

Scientific notation: `1.3339158110547646e-22`  
Decimal form: `0.0000000000000000000001333915811`

When pvalue is less than alpha level, we reject the Ho, it will be Ha.

When pvalue is greater than alpha level, we fail to reject the Ho.

#  T-test: Statistic and P-value

---

## 1. T-Statistic (Test Statistic)

###  Definition:
T-statistic is a value which tells about **the sample mean, how far away from population mean**, in terms of standard error.

###  Formula (1-sample t-test):

\$[
t = \frac{\bar{x} - \mu}{s / \sqrt{n}}
\$]

###  Where:

- \$(\bar{x}\$) = Sample mean  
- \$(\mu\$) = Population mean (from H₀)  
- \$(s\$) = Sample standard deviation  
- \$(n\$) = Sample size  

###  Interpretation:

- If the value of `t` **very bigger or very smaller** (positive or negative), this means:
  > The sample result is **very different** from expected value.

---

##  2. P-value

###  Definition:
P-value is **probability** which tells about:

> "**There is a chance of getting the result (or very extreme result)**, if Null Hypothesis (H₀) is TRUE?"

###  Interpretation:

| P-value         | Decision about H₀              |
|------------------|-------------------------------|
| p > 0.05         | Accept (Fail to reject) H₀     |
| p ≤ 0.05         | Reject H₀                      |
| p ≈ 0            | Strong evidence against H₀     |

- **Low p-value (e.g., 0.01)** → Null hypothesis could be FALSE.
- **High p-value (e.g., 0.5)** → The difference value could be by chance.

---

##  Summary Table

| Term         | Meaning                                                  | Formula / Logic                                   |
|--------------|----------------------------------------------------------|---------------------------------------------------|
| T-statistic  | How far the result is from the expected value (in SE units)    | \$(\frac{\bar{x} - \mu}{s / \sqrt{n}}\$)            |
| P-value      | The chances of the result if Null hypothesis is TRUE | Probability based on t-distribution curve         |

---


There You Can see in pvalue =e-22 means, before starting 1.339 there are 22 times 0

Lets Write in this way
When you run ttest , we get 2 values 1) t statistics score, 2) p value for two tail.

**pvalue /2**

In [20]:
t,p=st.ttest_1samp(hyp.Math1,mu     # Saving probability value in p and statistics value in t

In [22]:
t

-11.102736525713398

In [24]:
p

1.3339158110547646e-22

#  2 Sample Tests 

When you want to compare **two averages**, you use a **2-sample t-test**.  
There are **two types**: Independent and Paired.

---

## A. Independent Sample T-Test

###  What is it?

Used when you are comparing the **average of two separate groups** — like comparing boys and girls.

> The two groups **must not overlap**.

---

###  Example:

**Compare Math1 scores of Males vs. Females**

In the dataset:

- `Gender == 1` → Male  
- `Gender == 0` → Female  
- `Math1` → First math test score

```python
import scipy.stats as st

# Perform Independent Sample T-Test
t, p = st.ttest_ind(hyp[hyp['Gender'] == 1].Math1,
                    hyp[hyp['Gender'] == 0].Math1)

print("T-statistic:", t)
print("P-value:", p)


###  Interpretation

- **T-statistic** tells you **how different the two groups are**.
- **P-value** tells you **whether that difference is statistically significant**.

---

###  If `p < 0.05`:

> "Yes! The difference is significant."  
> For example:  
> "Males and females have significantly different Math1 scores."

---

###  If `p >= 0.05`:

> "No, the difference is not significant."  
> For example:  
> "There is no strong evidence that males and females have different Math1 scores."


# A- Independent Sample

In [30]:
hyp.head()

ID  Gender  Race  SEB  School  Prog  Read  Write  Math1  Math2  SST
0   70       0     4    1       1     1    57     52     41     47   57
1  121       1     4    2       1     3    68     59     53     63   61
2   86       0     4    3       1     1    44     33     54     58   31
3  141       0     4    3       1     3    63     44     47     53   56
4  172       0     4    2       1     2    47     52     57     53   61

In [32]:
hyp['Gender'].value_counts()

Gender
1    109
0     91
Name: count, dtype: int64

In [34]:
#How to Subset the Data on Gender

hyp['Gender']==0

0       True
1      False
2       True
3       True
4       True
       ...  
195    False
196    False
197    False
198    False
199    False
Name: Gender, Length: 200, dtype: bool

In [36]:
# To get the data form with the results for male only
hyp[hyp['Gender']==0]

ID  Gender  Race  SEB  School  Prog  Read  Write  Math1  Math2  SST
0    70       0     4    1       1     1    57     52     41     47   57
2    86       0     4    3       1     1    44     33     54     58   31
3   141       0     4    3       1     3    63     44     47     53   56
4   172       0     4    2       1     2    47     52     57     53   61
5   113       0     4    2       1     2    44     52     51     63   61
..  ...     ...   ...  ...     ...   ...   ...    ...    ...    ...  ...
87  117       0     4    3       1     3    34     49     39     42   56
88  133       0     4    2       1     3    50     31     40     34   31
89   94       0     4    3       1     2    55     49     61     61   56
90   24       0     2    2       1     2    52     62     66     47   46
91  149       0     4    1       1     1    63     49     49     66   46

[91 rows x 11 columns]

In [38]:
hyp[hyp['Gender']==1].Math1.head()

1     53
92    65
93    52
94    46
95    61
Name: Math1, dtype: int64

In [40]:
t,p=st.ttest_ind(hyp[hyp['Gender']==1].Math1,hyp[hyp['Gender']==0].Math1)

In [42]:
t

-0.41299864929688507

In [44]:
p

0.6800544974232143

#  Paired Sample T-Test

Used when you are **comparing two related measurements** from the **same individuals**  — like **Math1 vs. Math2** scores of the same students, or **before and after** a training program. Paired sample T-test is also called as relational test.

Each person contributes **both values**.

In [47]:
t,p=st.ttest_rel(hyp.Math1,hyp.Math2)

In [49]:
t

1.3555927189268238

In [51]:
p

0.1767652560550735

###  What if We Have More Than Two Samples?

The **1-sample t-test** is used when we have **only one sample** and we want to compare it with a known population mean.

The **2-sample t-test** is used to compare the means of **two groups**.

But what if we have **more than two groups or samples**?

 In that case, using multiple t-tests increases the chance of making a mistake (Type I error).  
So, instead of running many t-tests, we use **ANOVA (Analysis of Variance)** — a statistical method designed to compare **more than two groups at the same time**.

# 2 ANOVA (Analysis of Variance)

ANOVA is used to **compare the means of more than two groups**.

It works by analyzing the **variation in the data** and identifying where the variation comes from — either **between groups** or **within groups**.  
That’s why it’s called **Analysis of Variance**.

---

##  How ANOVA Works

ANOVA compares:
- The **variation between group means**, and  
- The **variation within each group**.

---

##  Assumptions:

- **Null Hypothesis (H₀)**: All group means are equal.  
- **Independent variables** are **categorical** (e.g., Gender, Program Type).  
- **Dependent variable** is **continuous** (e.g., Test Scores, Sales, Height).

---

##  F-value

The **F-value** helps us understand the ratio of variation:

\$[
F = \frac{\text{Variation Between Groups}}{\text{Variation Within Groups}}
\$]

- A **higher F-value** suggests a **greater difference between group means**.
- Based on the **F-score**, we calculate the **p-value** to check significance.
- If the **F-value is high**, the **null hypothesis is likely to be rejected**, meaning **the group means are significantly different**.
- If the **F-value is low**, it suggests that the **null hypothesis is correct**, i.e., **group means are not significantly different**.

---

> **"F"** in **F-score** stands for **Fisher**.  
> The F-score is named after **Sir Ronald Fisher**, a famous statistician who developed this method.

---

## Example

If different types of drinks (coffee, water, milk, soda, etc.) have the same effects on the human body (e.g., reaction time in the morning).

Mean Reaction time

Null Hypothesis: The average reaction time on the human body is the same for all types of drinks.

Alternate Hypothesis: The average reaction time on the human body varies among different types of drinks.

# Hypothesis Construction

Null Hypothesis for ANOVA states that the averages of dependent variables are the same across all the given groups.

The alternative hypothesis states that the means of the dependent variables are not the same for the given groups.

Ho:muA=muB=muC

Ha:not all are equal

# How its work behind the Scene?

**step1:** Calculate the mean of each semester

**step2:** Calculate the grand mean

**step3:** Variation between,within

**step4:** calculate the mean squared variance of between and within

**step5:** calculate F statistics and corresponding p value

In anova test first we have to find F Score. Then we have to find Pvalues. Ultimately we have to find P values

In [6]:
import pandas as pd
hyp=pd.read_csv('https://raw.githubusercontent.com/s4sauravv/Datasets/main/Semester%20marks.csv')

In [61]:
hyp

Student_ID  Semester  Total_Marks
0             1         1          365
1             2         3          376
2             3         1          411
3             4         3          436
4             5         2          366
..          ...       ...          ...
195         196         1          431
196         197         3          330
197         198         1          311
198         199         1          358
199         200         2          468

[200 rows x 3 columns]

--> As per Annova we need to find continous data based on the categorical variable 

- Total_Marks is continue variable and

- and Semester are categorical variable

**Ho** - Average Marks of each semester are almost equal

**Ha** - is all are not equal

In [65]:
import numpy as np
import statsmodels.formula.api as sm
mod=sm.ols('Total_Marks ~ Semester',data=hyp).fit()

fit() -When we use fit?

The fit() is started finding the relationship between semester 1,semester2 ,semester 3. it will start learning ,How is Variance and trying to find pattern, model is learning, and learn as much as possible. 

ols= Ordinary least square method

tilde "~"

lm -Linear Model

In [72]:
import statsmodels.api as k
avo_table=k.stats.anova_lm(mod,typ=1)
avo_table

df         sum_sq      mean_sq         F    PR(>F)
Semester    1.0     291.846746   291.846746  0.102572  0.749102
Residual  198.0  563366.873254  2845.287239       NaN       NaN

In [8]:
# Lets see manually if the average marks of each semester are close to each other 

#Semester 1 Average Marks
import numpy as np
print("1st Sem Avg Marks =  ",np.ceil(hyp['Total_Marks'][hyp['Semester']==1].mean()))
print("2nd Sem Avg Mraks =  ",np.ceil(hyp['Total_Marks'][hyp['Semester']==2].mean()))
print("3rd Sem Avg Marks =  ",np.ceil(hyp['Total_Marks'][hyp['Semester']==3].mean()))

1st Sem Avg Marks =   403.0
2nd Sem Avg Mraks =   393.0
3rd Sem Avg Marks =   406.0


# 3 Chi- Sample Test

The test most of the used for Goodness of Fit to a distribution  : -  The chi squared test can be used to determine wheather your data obeys a known theoretical porbablity distribution such as normal or poisson distribution.

Example - How Much sales we can expect expect this week (lets say pizza business).According to a salesman below are the details

#### Expected Sales in Thousand ..........[25,28,21,30,45,58,57]

#### Observed(Actual) Sales in thousand ..............[23,26,25,31,48,63,58]

How expected and observed close to each other, how good it is, this is a goodness of fit.

# Lets understand how to work on this situation.

We will do this hypothesis testing with a significance of 10% (0.1) (This will be given or you can test accordindg to business intrect)

![ChiSquare](chisq.png)

## Chi-Square Statistics Formula

In [2]:
import numpy as np

#Expected [25,28,21,30,45,58,57]
#Observer [23,26,25,31,48,63,58]

chi_ind=[np.square(23-25)/25,
        np.square(26-28)/25,
        np.square(25-21)/21,
        np.square(31-30)/30,
        np.square(48-45)/45,
        np.square(63-58)/58,
        np.square(58-57)/57]

print("Chi-Square test statistics score =",np.sum(chi_ind))

Chi-Square test statistics score = 1.763816437645839


# Lets Check Square Distribution table

### What is the Chi-Square Distribution table ?

The Chi-Sqaure dsitribution table is a table that shows the critical values of the chi-square distribution.To use the Chi-Square Distribution table, you only need to know two values.

--> The Degrees of Freedom for the chi-square test
--> The alpha level for the test (common choices are 0.01,0.05,0.10)

**Degree of Freedom** - Which means you are allowed to choose/vary the data at the certain extent, means. Freedom is given to vary the data example -(N-1).Here N is Total events i.e(7-1)=6

DF = 6 and Significance level =0.10 and our test score is = 2.58

#--> lets check the critical score where DF and Significance level converge

![ftable](ftable.png)

# Conclusion -- 
 As the Critical Score (in the table )is greater than our test score (1.76) , we cannot reject the       null hypothesis which also means that sales man prediction/expectation was correct.

#  Covariance

--> A Systematic relationship between a pair of random variables where in a change in one variable reciprocated by an equivalent cahnge in another variable .

--> Covariance can take any value between -infinity to +infinity , where in the negative value is an indicator relationship wheareas a positive values represents the positive . relationship and when the value is zero, it indicates no relationship.

--> In addition to this ,when all the observation of the either variable are same ,the covariance will be zero.

--> When we change the unit of observation or any or both the two variables ,then there is no change in the strength of the relationship between two varibles but the value of covariance in changed.

![Covairance](cov.png)

# Corelation

###  Formula:

\$[
\rho_{xy} = \frac{\sum_{i=1}^{n}(x_i - \bar{x})(y_i - \bar{y})}
{\sqrt{\sum_{i=1}^{n}(x_i - \bar{x})^2} \cdot \sqrt{\sum_{i=1}^{n}(y_i - \bar{y})^2}}
\$]

---
###  Explanation of Terms:

- \$( x_i \$): i-th value of variable x  
- \$( y_i \$): i-th value of variable y  
- \$( \bar{x} \$): Mean of x  
- \$( \bar{y} \$): Mean of y  
- \$( n \$): Total number of data points  

---

--> A Measure which determines the change in one variable dure to change in ohter variable .

-->Corelation is of two types  .i.e positive corelation or negative 

--> corelation can take any value between -1 to +1  represent strong positive corellation and values close to -1 is an indicator and values close to -1 is an indicator of strong negative corelation

--> There are four measure of corelation

- Scatter Diagram

- Product moment corellation coefficient

- Rank corellation coefficient 

- Coefficient of concurrent deviation

# Difference between Corelation and Covariance

![Difference](diff.png)